geoana/gravity.py

```diff
@@ -0,0 +1,410 @@
+        >>> ax6.contourf(X, Y, g_tens[:,:,1,2])
+        >>> ax7.contourf(X, Y, g_tens[:,:,2,0])
+        >>> ax8.contourf(X, Y, g_tens[:,:,2,1])
+        >>> ax9.contourf(X, Y, g_tens[:,:,2,2])
+        >>> plt.show()
+        """
+
+        r_vec = xyz - self.location
+        r = np.linalg.norm(r_vec, axis=-1)
+        g_tens = -G * self.mass * (np.eye(3) / r[..., None, None] ** 3 -
+                                   3 * r_vec[..., None] * r_vec[..., None, :] / r[..., None, None] ** 5)
+        return g_tens
+
+
+class Sphere(PointMass):
+    """Class for gravitational solutions for a sphere.
+
+    The ``Sphere`` class is used to analytically compute the gravitational
+    potentials, fields, and gradients for a sphere.
+
+    Parameters
+    ----------
+    density : float
+        Density of sphere (m).  Default is r = 1 kg/m^3.
+    radius : float
+        Radius of sphere (m).  Default is r = 1 m.
+    mass : float
+        Mass of the sphere (kg). Default is m = 4/3*pi*R^3*rho kg.
+    location : array_like, optional
+        Center of the sphere (m). Default is (0, 0, 0).
+    """
+
+    def __init__(self, density=1.0, radius=1.0, mass=1.0, location=None, **kwargs):
+        super().__init__(mass, location, **kwargs)
+        self.radius = radius
+        self.density = density
+
+    @property
+    def radius(self):
+        """Radius of the sphere in meters
+
+        Returns
+        -------
+        float
+            Radius of the sphere in meters
+        """
+        return self._radius
+
+    @radius.setter
+    def radius(self, item):
+        item = float(item)
+        if item < 0.0:
+            raise ValueError('radius must be non-negative')
+        self._radius = item
+
+    @property
+    def density(self):
+        """Density of the sphere in kilogram over meters cubed.
+
+        Returns
+        -------
+        float
+            Density of the sphere in kilogram over meters cubed.
+        """
+        return self.density
+
+    @density.setter
+    def density(self, item):
+        item = float(item)
+        if item < 0.0:
+            raise ValueError('density must be non-negative')
+        self.density = item
+
+    @PointMass.mass.getter
+    def mass(self):
+        return 4/3 * np.pi * self.radius ** 3 * self.density
+
+    def gravitational_potential(self, xyz):
+        """
+        Gravitational potential due to a sphere.
+
+        .. math::
+
+            U(P) = \\gamma \\frac{m}{r}
+
+        Parameters
+        ----------
+        xyz : (..., 3) numpy.ndarray
+            Locations to evaluate at in units m.
+
+        Returns
+        -------
+        (..., ) numpy.ndarray
+            Gravitational potential at sphere location xyz in units m^2/s^2.
+        """
+
+        r_vec = xyz - self.location
+        r = np.linalg.norm(r_vec, axis=-1)
+        u_g = np.zeros_like(r)
+        ind0 = r > self.radius
+        u_g[ind0] = PointMass.gravitational_potential(xyz)
```

Comment:
in newer versions of python, we can call the parent class's methods as `super().gravitational_potential(xyz)`